p4-16/spec/P4-16-spec.mdk

```diff
@@ -7097,16 +7100,157 @@ annotationBody
     | annotationBody annotationToken
 ~ End P4Grammar
 
-Annotations may impose additional structure on their bodies, and are not
-confined to the P4 language. For example, the P4Runtime specification defines a
-`@pkginfo` annotation that takes key-value pairs, which are not found in the
-main P4 language.
+Unstructured annotations may impose additional structure on their bodies, and
+are not confined to the P4 language. For example, the P4Runtime specification
+defines a `@pkginfo` annotation that takes key-value pairs, which are not found
+in the main P4 language.
+
+## Bodies of Structured Annotations
+
+Unlike unstructured annotations, structured annotations use square brackets
+`[...]` and have a restricted format. They are commonly used to declare custom
+metadata, consisting of expression lists or key-value lists but not both. An
+`expressionList` may be empty or contain a comma-separated list of member
+`expression`s. A `kvList` consists of one or more `kvPair`s, each consisting of
+a key and a value `expression`. Note the syntax for `expression` is rich, see
+Appendix [#sec-grammar] for details.
+
+To avoid ambiguity, an empty `structuredAnnotationBody` is by definition
+considered to contain an empty `expressionList`.
+
+All `expression`s within a `structuredAnnotationBody` must be compile-time
+known values, either literals or expressions requiring compile-time
+evaluation, with a result type that is one of: string literals,
+infinite-precision integer, or boolean. Nested `expression`s (e.g. an
+`expression` containing an `expressionList`, a `kvList`, etc.) are not
+allowed. Note that P4Runtime information (`P4Info`) may stipulate additional
+restrictions. For example, an integer expression might be limited to 64-bit
+values.
+
+It is illegal to duplicate a `key` within the `kvList` of a structured
+annotation.  It is illegal to use the same annotation `name` on a given
+`annotationToken` for structured annotations.
```

Comment:
As written, this sentence makes no sense: an annotation token is a part of the body of an unstructured annotation.

It should also be made more explicit that it is a program error to have a structured annotation and an unstructured annotation with the same name on a table, control, etc. I interpreted the current wording as just saying you can't have two structured annotations with the same name.